core/server/worker/src/test/java/alluxio/worker/block/DefaultBlockWorkerTest.java

```diff
@@ -297,4 +445,208 @@ public void getBlockReader() throws Exception {
+  }
+
+  @Test
+  public void cleanUpSession() throws Exception {
+    long sessionId = mRandom.nextLong();
+    long blockId = mRandom.nextLong();
+
+    mBlockWorker.createBlock(
+        sessionId,
+        blockId,
+        0,
+        new CreateBlockOptions(null, Constants.MEDIUM_MEM, 1));
+    mBlockWorker.commitBlock(sessionId, blockId, false);
+
+    // just to hold a read lock on the block
+    BlockReader reader = mBlockWorker.createBlockReader(sessionId, blockId, 0, false, null);
+
+    long anotherSessionId = mRandom.nextLong();
+
+    // clean up the first session
+    mBlockWorker.cleanupSession(sessionId);
+
+    // now another session should be able to grab write lock on the block
+    mBlockWorker.removeBlock(anotherSessionId, blockId);
+  }
+
+  private void cacheBlock(boolean async) throws Exception {
+    // flush 1MB random data to ufs so that caching will take a while
+    long ufsBlockSize = 1024 * 1024;
+    byte[] data = new byte[(int) ufsBlockSize];
+    mRandom.nextBytes(data);
+
+    try (FileOutputStream fileOut = new FileOutputStream(mTestUfsFile);
+         BufferedOutputStream bufOut = new BufferedOutputStream(fileOut)) {
+      bufOut.write(data);
+      bufOut.flush();
+    }
+
+    // ufs options: delegate to the ufs mounted at UFS_MOUNT_ID
+    // with path to our test file
+    long blockId = mRandom.nextLong();
+    Protocol.OpenUfsBlockOptions options = Protocol.OpenUfsBlockOptions
+        .newBuilder()
+        .setBlockSize(ufsBlockSize)
+        .setUfsPath(mTestUfsFile.getAbsolutePath())
+        .setMountId(UFS_MOUNT_ID)
+        .setNoCache(false)
+        .setOffsetInFile(0)
+        .build();
+
+    // cache request:
+    // delegate to local ufs client rather than remote worker
+    CacheRequest request = CacheRequest
+        .newBuilder()
+        .setSourceHost(NetworkAddressUtils.getLocalHostName(500))
+        .setBlockId(blockId)
+        .setLength(ufsBlockSize)
+        .setAsync(async)
+        .setOpenUfsBlockOptions(options)
+        .build();
+
+    mBlockWorker.cache(request);
+
+    // check that the block metadata is present
+    if (async) {
+      assertFalse(mBlockWorker.getBlockStore().hasBlockMeta(blockId));
+      waitFor(
+          "Wait for async cache",
+          () -> mBlockWorker.getBlockStore().hasBlockMeta(blockId),
+          WaitForOptions.defaults().setInterval(10).setTimeoutMs(2000));
+    } else {
+      assertTrue(mBlockWorker.getBlockStore().hasBlockMeta(blockId));
+    }
+
+    long sessionId = mRandom.nextLong();
+    // check that we can read the block locally
+    try (BlockReader reader = mBlockWorker.createBlockReader(
+            sessionId, blockId, 0, false, null)) {
+      ByteBuffer buf = reader.read(0, ufsBlockSize);
+      // alert: LocalFileBlockReader uses a MappedByteBuffer, which does not
+      // support the array operation. So we need to compare ByteBuffer manually
+      assertEquals(0, buf.compareTo(ByteBuffer.wrap(data)));
+    }
+  }
```

Comment:
The OpenUfsOptions is set to be null to avoid fallback behavior, so that the block has to be present locally.